algorithms/src/snark/varuna/ahp/prover/round_functions/fourth.rs

```diff
@@ -146,35 +151,36 @@ impl<F: PrimeField, MM: SNARKMode> AHPForR1CS<F, MM> {
 
     #[allow(clippy::too_many_arguments)]
     fn calculate_matrix_sumcheck_witness(
-        label: String,
+        label: &str,
+        id: CircuitId,
         constraint_domain: EvaluationDomain<F>,
         variable_domain: EvaluationDomain<F>,
         non_zero_domain: EvaluationDomain<F>,
-        arithmetization: &MatrixArithmetization<F>,
+        arithmetization: &MatrixEvals<F>,
         alpha: F,
         beta: F,
         v_R_i_alpha_v_C_i_beta: F,
         max_non_zero_domain: EvaluationDomain<F>,
         fft_precomputation: &FFTPrecomputation<F>,
         ifft_precomputation: &IFFTPrecomputation<F>,
-    ) -> Result<(Sum<F>, Lhs<F>, Gpoly<F>)> {
+    ) -> Result<(Sum<F>, Lhs<F>, Gpoly<F>, Apoly<F>, Bpoly<F>)> {
+        let (row_on_K, col_on_K, row_col_val) =
+            (&arithmetization.row, &arithmetization.col, &arithmetization.row_col_val);
+        let R_size = constraint_domain.size_as_field_element;
+        let C_size = variable_domain.size_as_field_element;
+
         let mut job_pool = snarkvm_utilities::ExecutionPool::with_capacity(2);
         job_pool.add_job(|| {
             let a_poly_time = start_timer!(|| format!("Computing a poly for {label}"));
             let a_poly = {
-                let coeffs = cfg_iter!(arithmetization.row_col_val.as_dense().unwrap().coeffs())
-                    .map(|v| v_R_i_alpha_v_C_i_beta * v)
-                    .collect();
-                DensePolynomial::from_coefficients_vec(coeffs)
+                let evals = cfg_iter!(row_col_val.evaluations).map(|v| v_R_i_alpha_v_C_i_beta * v).collect();
+                EvaluationsOnDomain::from_vec_and_domain(evals, non_zero_domain)
+                    .interpolate_with_pc(ifft_precomputation)
```

Comment:
Hm looks like we're adding an extra IFFT here now, which is non-ideal...

If you have the time, it would be good to figure out the impact of this on the prover time

Reply:
Indeed, however fortunately overall time for a typical `test_transfer_public_transaction_size` even goes *down*. One sample (I ran multiple):
- proving time before: 98.84s user 12.05s system 766% cpu 14.475 total
- proving time after: 96.65s user 10.57s system 772% cpu 13.881 total

I think the reduction may come from the fact that the prover is aggregating a smaller number of polynomials during opening (from 4 (row,col,rowcol,rowcolval) to 2 (a,b) indexing polynomials per circuit)

Reply:
Lovely! (I imagine the commitment aggregation step is dwarfed by the smaller amount of work performed. I wonder if there's similar redundancies elsewhere. E.g., could we apply the same idea to the first sumcheck? We're already computing the random linear combinations of the polynomials there, so we shouldn't have to repeat work *again* inside the PC scheme.

Reply:
Yes interesting thought. But I don't see obvious contenders. Will make a TODO to investigate if we can re-use the entire "left hand side" in the lineval sumcheck.

Reply:
I think the savings would happen here. 

In `construct_linear_combinations`, we take a linear combination of the `w_j`s: https://github.com/AleoHQ/snarkVM/blob/b87eb05a93703ddb8a9b4bff53a5d129d92741a2/algorithms/src/snark/varuna/ahp/ahp.rs#L298. This is called when we invoke `SonicPC::open`.

Meanwhile, in https://github.com/AleoHQ/snarkVM/blob/b87eb05a93703ddb8a9b4bff53a5d129d92741a2/algorithms/src/snark/varuna/ahp/prover/round_functions/third.rs#L313, we compute z_m, but we can already multiply in the random linear combiners into `assignment` before hand, and save this result.

(Also, separately, I think in the latter file we don't need to compute the individual `g_i`s and `h_i`s; we should just directly compute `g_1` and `h_1` by multiply the `z_m_i` with the vanishing poly for the `column_i` domain.)

Overall I suspect these changes should knock off another 5-10% off the time.

Reply:
Cool, thx for sharing the details, appreciate it. Will tackle in a next PR!

Reply:
Let's extract this off into another issue (it's just a prover-side improvement, and so can be done at a later stage)